src/vconsole/vconsole-setup.c

```diff
@@ -54,6 +54,55 @@ static bool is_vconsole(int fd) {
         return ioctl(fd, TIOCLINUX, data) >= 0;
 }
 
+static bool is_allocated(int fd, int idx) {
+        struct vt_stat vcs = {};
+        char vcname[strlen("/dev/vcs") + DECIMAL_STR_MAX(int)];
+
+        if (fd >= 0) {
+                if (ioctl(fd, VT_GETSTATE, &vcs) < 0) {
+                        (void) log_warning_errno(errno, "VT_GETSTATE failed: %m");
+                        return false;
+                }
+                idx = vcs.v_active;
+        }
+
+        xsprintf(vcname, "/dev/vcs%i", idx);
+        if (access(vcname, F_OK) < 0)
+                return false;
+        return true;
+}
+
+static bool is_settable(int *fd, int idx) {
```

Comment:
So... after reading the whole series, I think that this `is_settable` interface is a source of much of the complications later on. I think you should make this function simple, by requiring a valid fd as an argument, and add a wrapper function `open_index_for_setting` or similar with a better name:
```
static int open_index_for_setting(int idx) {
                  xsprintf(vcname, "/dev/tty%i", idx);
                  fd = open_terminal(vcname, O_RDWR|O_CLOEXEC);
                  if (fd < 0)
                          return fd;
                  if (!is_settable(fd)) {
                           close(fd);
                          return -EBUSY;
                  }
                  return fd;
}
```